src/iceberg/avro/avro_schema_util.cc

```diff
@@ -783,4 +785,201 @@ Result<SchemaProjection> Project(const Schema& expected_schema,
+  }
+}
+
+Result<::avro::NodePtr> CreateRecordNodeWithFieldIds(const ::avro::NodePtr& original_node,
+                                                     const NameMapping& name_mapping) {
+  auto new_record_node = std::make_shared<::avro::NodeRecord>();
+  new_record_node->setName(original_node->name());
+
+  for (size_t i = 0; i < original_node->leaves(); ++i) {
+    const std::string& field_name = original_node->nameAt(i);
+    ::avro::NodePtr field_node = original_node->leafAt(i);
+
+    // Try to find field ID by name in the name mapping
+    if (auto field_ref = name_mapping.Find(field_name)) {
+      if (field_ref->get().field_id.has_value()) {
+        // Add field ID attribute to the new node
+        ::avro::CustomAttributes attributes;
+        attributes.addAttribute(std::string(kFieldIdProp),
+                                std::to_string(field_ref->get().field_id.value()), false);
+        new_record_node->addCustomAttributesForField(attributes);
+      }
+
+      // Recursively apply field IDs to nested fields if they exist
+      if (field_ref->get().nested_mapping && field_node->type() == ::avro::AVRO_RECORD) {
+        const auto& nested_mapping = field_ref->get().nested_mapping;
+        auto fields_span = nested_mapping->fields();
+        std::vector<MappedField> fields_vector(fields_span.begin(), fields_span.end());
+        auto nested_name_mapping = NameMapping::Make(std::move(fields_vector));
+
+        ICEBERG_ASSIGN_OR_RAISE(
+            auto new_nested_node,
+            CreateAvroNodeWithFieldIds(field_node, *nested_name_mapping));
+        new_record_node->addName(field_name);
+        new_record_node->addLeaf(new_nested_node);
+      } else {
+        // Recursively apply field IDs to child nodes
+        ICEBERG_ASSIGN_OR_RAISE(auto new_field_node,
+                                CreateAvroNodeWithFieldIds(field_node, name_mapping));
+        new_record_node->addName(field_name);
+        new_record_node->addLeaf(new_field_node);
+      }
+    } else {
+      // Recursively apply field IDs to child nodes even if no mapping found
+      ICEBERG_ASSIGN_OR_RAISE(auto new_field_node,
+                              CreateAvroNodeWithFieldIds(field_node, name_mapping));
+      new_record_node->addName(field_name);
+      new_record_node->addLeaf(new_field_node);
+    }
+  }
+
+  return new_record_node;
+}
+
+Result<::avro::NodePtr> CreateArrayNodeWithFieldIds(const ::avro::NodePtr& original_node,
+                                                    const NameMapping& name_mapping) {
+  if (original_node->leaves() != 1) {
+    return InvalidSchema("Array type must have exactly one leaf");
+  }
+
+  auto new_array_node = std::make_shared<::avro::NodeArray>();
+  new_array_node->setName(original_node->name());
+  new_array_node->setLogicalType(original_node->logicalType());
+
+  // Check if this is a map represented as array
+  if (original_node->logicalType().type() == ::avro::LogicalType::CUSTOM &&
```

Comment:
Please reuse `bool HasMapLogicalType(const ::avro::NodePtr& node)` in the `avro_schema_util_internal.h`